modules/apps/callbacks/ibc_middleware.go

```diff
@@ -0,0 +1,309 @@
+	}
+
+	err := im.processCallback(ctx, packet, types.CallbackTypeWriteAcknowledgement, callbackDataGetter, callbackExecutor)
+	if err != nil {
+		// revert entire tx if processCallback returns an error
+		panic(err)
+	}
+
+	return ack
+}
+
+// WriteAcknowledgement implements the WriteAcknowledgement destination callbacks for the ibc-callbacks middleware during
+// asynchronous packet acknowledgement.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback fails (within the gas limit), state changes are reverted.
+func (im IBCMiddleware) WriteAcknowledgement(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	packet ibcexported.PacketI,
+	ack ibcexported.Acknowledgement,
+) error {
+	err := im.ics4Wrapper.WriteAcknowledgement(ctx, chanCap, packet, ack)
+	if err != nil {
+		return err
+	}
+
+	packetReceiverAddress := im.GetPacketReceiver(packet)
+	callbackDataGetter := func() (types.CallbackData, bool, error) {
+		return types.GetDestCallbackData(im.app, packet.GetData(), ctx.GasMeter().GasRemaining(), im.maxCallbackGas)
+	}
+	callbackExecutor := func(cachedCtx sdk.Context, callbackAddress string) error {
+		return im.contractKeeper.IBCWriteAcknowledgementCallback(cachedCtx, packet, ack, callbackAddress, packetReceiverAddress)
+	}
+
+	return im.processCallback(ctx, packet, types.CallbackTypeWriteAcknowledgement, callbackDataGetter, callbackExecutor)
+}
+
+// SendPacket implements the ICS4 Wrapper interface
+func (im IBCMiddleware) SendPacket(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	sourcePort string,
+	sourceChannel string,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	data []byte,
+) (uint64, error) {
+	seq, err := im.ics4Wrapper.SendPacket(ctx, chanCap, sourcePort, sourceChannel, timeoutHeight, timeoutTimestamp, data)
+	if err != nil {
+		return seq, err
+	}
+
+	// we use the reconstructed packet to get the packet sender, this should be fine since the only missing fields are
+	// the destination port and channel. And GetPacketSender is a static method that does not depend on the context, so
+	// it should be fine to use the reconstructed packet.
+	reconstructedPacket := channeltypes.NewPacket(data, seq, sourcePort, sourceChannel, "", "", timeoutHeight, timeoutTimestamp)
+	packetSenderAddress := im.GetPacketSender(reconstructedPacket)
+
+	callbackDataGetter := func() (types.CallbackData, bool, error) {
+		return types.GetSourceCallbackData(im.app, data, ctx.GasMeter().GasRemaining(), im.maxCallbackGas)
+	}
+	callbackExecutor := func(cachedCtx sdk.Context, callbackAddress string) error {
+		return im.contractKeeper.IBCSendPacketCallback(
+			cachedCtx, sourcePort, sourceChannel, timeoutHeight, timeoutTimestamp, data, callbackAddress, packetSenderAddress,
+		)
+	}
+
+	return seq, im.processCallback(ctx, reconstructedPacket, types.CallbackTypeSendPacket, callbackDataGetter, callbackExecutor)
```

Comment:
I didn't intend to give `SendPacket` the ability to block a send. I thought the contract should have no say in this because if contract is the sender, then it can do the validation before sending, and if the contract is not, I don't think it should be able to block someone else's packet. I thought this callback was more the contract preparing itself for a future callback or subscribing to the callback or something like this. 

But I can also see the use case for what you mentioned. I'll bring it up in our group chat

`processCallback` could be reused even in this case. We could allow a specific type of error that should be returned to block a packet send.